Fixing BLE connect logic

massi-ang opened the pull request:
*Issue #, if available:*

*Description of changes:*
Fixing the way the the SDK manages BLE connect logic on Android based on information from 
https://medium.com/@martijn.van.welie/making-android-ble-work-part-2-47a3cdaade07

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -948,6 +953,7 @@ private void setMtu(int mtu) {
         if (mBleConnectionState == BleConnectionState.BLE_CONNECTED && mBluetoothGatt != null) {
             Log.i(TAG, "Setting mtu to: " + mtu);
             sendBleCommand(new BleCommand(REQUEST_MTU, mtu));
+            mBleConnectionStatusCallback.onBleConnectionStatusChanged(mBleConnectionState);
```
why is the connection status change callback required here?

massi-ang replied:
BLE_CONNECTED is not useful to the application, since what is important is that the GATT service has been initialized. This is the reason to put the callback here.
It is probably better to add a new state, `BLE_INITIALIZED` or similar, which should notify the app that the BLE services have been initialized and are ready to be used. I would set that state in the `onMtuChanged` callback, and also call the application connection callback handler there.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -168,10 +169,12 @@ void disconnect() {
-
+        /**
+         * Call disconnect but not close. Close will be called in the Gatt callback.
+         * This allows Android to track the changes
```
what happens if GATT is already disconnected and user call disconnect ? Will the callback be invoked with an error ?

massi-ang replied:
I now moved the whole disconnect logic in the onConnectionStatusChange() callback

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -185,9 +188,11 @@ void disconnect() {
         if (mMqttConnectionState != AmazonFreeRTOSConstants.MqttConnectionState.MQTT_Disconnected) {
             disconnectFromIot();
         }
+        mBleConnectionStatusCallback.onBleConnectionStatusChanged(mBleConnectionState);
```
Will this lead to callback invoked twice for a successful disconnect, other one being from onConnectionStateChange() ?

massi-ang replied:
I moved the whole disconnect logic to the onConnectionStateChange() callback.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -437,7 +447,7 @@ public void onServicesDiscovered(BluetoothGatt gatt, int status) {
                                 +mBluetoothDevice.getBondState());
                         describeGattServices(mBluetoothGatt.getServices());
-                        if (mBluetoothDevice.getBondState() == BluetoothDevice.BOND_BONDED) {
+                        if (mBluetoothDevice.getBondState() == BluetoothDevice.BOND_BONDING) {
```
Why is this within a bonding state ? From `onConnectionStateChange()` code it looks like `discoverServices()` will only be called when `bondState != BluetoothDevice.BOND_BONDING`

massi-ang replied:
Right, if we have successfully discovered the service, the bond_state should be correct so we can call initialize() without further check.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -805,21 +814,17 @@ private void subscribeToIoT(final Subscribe subscribe) {
-                                    data
-                            );
-                            publishToDevice(publish);
-                        } catch (UnsupportedEncodingException e) {
```
was this exception removed intentionally?

massi-ang replied:
Yes, since I have also changed this String message = new String(data, StandardCharsets.UTF_8); to use the constant and not an arbitrary string. In this was the method does not throw any exceptions anymore

massi-ang commented:
@ravibhagavandas I have done few fixes addressing your comments. 
Basically, after the connection to the BLE Gatt server is established, I perform a service discovery. When the service discovery is done I will try to read some characteristics (in `initialize()`), but not send any other command. In case bonding has been enabled on FreeRTOS this triggers the pairing. The remaining initialization is done when a characteristic is first read successfully, which copes with both bonding enabled and disabled cases.

massi-ang commented:
Devil's in the details. I need to read one single characteristic at first, otherwise if the pairing is cancelled, you get additional paring request for the other reads.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -490,10 +518,17 @@ public void onCharacteristicRead(BluetoothGatt gatt,
+                        // On the first successful read we enable the services
+                        if (mBleConnectionState == BleConnectionState.BLE_CONNECTED) {
+                            Log.d(TAG, "GATT services initialized");
+                            mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_INITIALIZED;
+                            mBleConnectionStatusCallback.onBleConnectionStatusChanged(mBleConnectionState);
```
I think the BLE_INITIALIZED callback should be invoked only after initialization step is completed, probably in `onMtuChanged` callback as it was done before.

massi-ang replied:
Will that call back always be invoked?

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSConstants.java:
```diff
@@ -69,10 +69,11 @@
+        BLE_CONNECTING,    // = 1
+        BLE_CONNECTED,     // = 2
+        BLE_DISCONNECTING, // = 3
+        BLE_INITIALIZED    // = 4
```
New state needs to be handled in the demo app in `DeviceScanFragment.java`.

massi-ang commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -536,6 +571,7 @@ public void onCharacteristicRead(BluetoothGatt gatt,
                                 try {
                                     mMtu = Integer.parseInt(currentMtu.mtu);
+                                    mMaxPayloadLen = mMtu - 3;
                                     if (mDeviceInfoCallback != null) {
```
setMtu() is called only if mDeviceInfoCallback is not null, hence if we move the setting to BLE_INITIALIZED state into onMtuChange, it might never be set.

massi-ang commented:
I have move the BLE_INITIALIZED state into the onMtuChanged callback and added a new BLE_INITIALIZING state to indicate that the initialization process has started.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -937,15 +941,15 @@ private void publishToDevice(final Publish publish) {
 
     private void setMtu(int mtu) {
-        if (mBleConnectionState == BleConnectionState.BLE_CONNECTED && mBluetoothGatt != null) {
+        if (mBleConnectionState == BleConnectionState.BLE_INITIALIZED && mBluetoothGatt != null) {
```
`setMtu()` is called while BLE is being initialized so this will skip setting MTU. probably check `isBLEConnected()` as it is valid as soon as BLE is connected.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -924,7 +928,7 @@ private void sendPubAck(final Publish publish) {
 
     private void publishToDevice(final Publish publish) {
-        if (mBleConnectionState != BleConnectionState.BLE_CONNECTED) {
+        if (!isBLEConnected()) {
```
check for `mBluetoothGatt != null` to be consistent with other methods ?

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -910,7 +914,7 @@ private void sendUnsubAck(final Unsubscribe unsubscribe) {
 
     private void sendPubAck(final Publish publish) {
-        if (mBleConnectionState != BleConnectionState.BLE_CONNECTED) {
+        if (!isBLEConnected()) {
```
check for mBluetoothGatt != null to be consistent with other methods ?

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -896,7 +900,7 @@ private void sendSubAck(final Subscribe subscribe) {
 
     private void sendUnsubAck(final Unsubscribe unsubscribe) {
-        if (mBleConnectionState != BleConnectionState.BLE_CONNECTED) {
+        if (!isBLEConnected()) {
```
check for mBluetoothGatt != null to be consistent with other methods ?

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -880,8 +879,13 @@ private void sendConnAck() {
+    }
     private void sendSubAck(final Subscribe subscribe) {
-        if (mBleConnectionState != BleConnectionState.BLE_CONNECTED) {
+        if (!isBLEConnected()) {
```
check for mBluetoothGatt != null to be consistent with other methods ?

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -478,9 +470,13 @@ public void onMtuChanged(BluetoothGatt gatt, int mtu, int status)  {
                     mMaxPayloadLen = mMtu - 3;
                     mMaxPayloadLen = mMaxPayloadLen > 0 ? mMaxPayloadLen : 0;
+                    // The BLE service should be initialized at this stage
+                    if ( mBleConnectionState == BleConnectionState.BLE_INITIALIZING) {
```
As  a side effect of this change `sendConnAck()`  at line 756 is failing since it checks for `BLE_CONNECTED` state before sending.Suggesting to move the check for BLE state inside `sendConnAck()` definition and check for `isBleConnected()` instead, as we do for other MQTT relay methods.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -409,24 +390,42 @@ public void onConnectionStateChange(BluetoothGatt gatt, int status,
+
+                            // If ble connection is closed, there's no need to keep mqtt connection open.
+                            //mBleConnectionState = AmazonFreeRTOSConstants.BleConnectionState.BLE_DISCONNECTED;
+                            if (mMqttConnectionState != AmazonFreeRTOSConstants.MqttConnectionState.MQTT_Disconnected) {
```
We need to call this in  `else` block as well when there is a disconnect due to error.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -409,24 +390,42 @@ public void onConnectionStateChange(BluetoothGatt gatt, int status,
-                        Log.i(TAG, "Disconnected from GATT server.");
-                        mBleConnectionStatusCallback.onBleConnectionStatusChanged(mBleConnectionState);
-                        //broadcastUpdate(intentAction);
+                        gatt.close();
```
The two lines needs to be wrapped in `if (!mGattAutoReconnect)` as this is also caused by peripheral initiated disconnect.

ravibhagavandas commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -409,24 +390,42 @@ public void onConnectionStateChange(BluetoothGatt gatt, int status,
-                        //broadcastUpdate(intentAction);
+                        gatt.close();
+                        mBluetoothGatt = null;
+                        disconnect();
```
call `cleanup()` here instead of `disconnect()`

dgranahan commented on amazonfreertossdk/src/main/java/software/amazon/freertos/amazonfreertossdk/AmazonFreeRTOSDevice.java:
```diff
@@ -536,11 +556,11 @@ public void onCharacteristicRead(BluetoothGatt gatt,
                                 Log.i(TAG, "Default MTU is set to: " + currentMtu.mtu);
                                 try {
                                     mMtu = Integer.parseInt(currentMtu.mtu);
+                                    mMaxPayloadLen = mMtu - 3;
```
Should this also do a check like line 480/489?
```
mMaxPayloadLen = mMaxPayloadLen > 0 ? mMaxPayloadLen : 0;
```

mangelino replied:
Can you do another PR to fix that?

dgranahan replied:
@massi-ang  https://github.com/aws/amazon-freertos-ble-android-sdk/pull/21
